client/rest/build.gradle

```diff
@@ -38,13 +61,66 @@ publishing {
   }
 }
 
+configurations {
+  shade {
+    transitive = false
+  }
+}
+
+// Useful for build time dependencies, as it is generated before compilation of the source in the rest client.
+// This cannot be used as the final shaded jar, as it will contain the compiled source and dependencies
+File shadedDir = file("${buildDir}/shaded")
+// This directory exists so that the shadeDeps task would produce an output, so we can add it (below) to the source set.
+File shadedSrcDir = file("${buildDir}/generated-dummy-shaded")
+task shadeDeps(type: com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar) {
+  destinationDir = shadedDir
+  configurations = [project.configurations.shade]
+  classifier = 'deps'
+  relocate 'org.apache', 'org.elasticsearch.client'
+
+  doLast {
+    shadedSrcDir.mkdir()
+  }
+}
+
+// builds the actual shaded jar that is being distributed
+shadowJar {
+  configurations = [project.configurations.shade]
+  classifier = null
+  relocate 'org.apache', 'org.elasticsearch.client'
```

Comment:
You are right, at this point I could combine the deps jar with the src/ in the project. I can follow up with this, as its a better idea than running another "very similar" shadow task. I can follow up with this